streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -894,6 +898,22 @@ public void run() {
         }
     }
 
+    /**
+     * Shutdown this {@code KafkaStreams} by signaling all the threads to stop, and then wait up to the timeout for the
+     * threads to join.
+     * A {@code timeout} of 0 means to wait forever.
```

Comment:
I'm fine with those semantics, maybe we can make a Jira.

Thinking about it more, though, wouldn't it make more sense to:
* reject negative numbers
* make 0 just signal and return immediately (after checking the state once)
* if I want to wait "forever", I can use `ofYears(1)` or `ofMillis(Long.MAX_VALUE)` or some other intuitively "long enough to be forever" value instead of a magic value.

Regardless, I agree the current behavior is a little weird, and I'd be in favor of a Jira/KIP to revise it.

Reply:
Created: https://issues.apache.org/jira/browse/KAFKA-7477

Reply:
I think, we should fix this in 2.1 -- I am suggestion this, because we could use the new semantics for `close(Duration)` only and stay with old semantics for `close(long, TimeUnit)` -- if we don't so this, it would be an backward incompatible change and thus we could only do it in `3.0.0.`.

Reply:
@nizhikov Are you willing for address 7477 in this PR? If not, it's also fine and we do a follow up PR. However, it should be part of the KIP description. Could you update the KIP accordingly?

Reply:
@mjsax I'll take care of KAFKA-7477 in follow up PR.

Reply:
Did you start working on KAFKA-7477 already @nizhikov ?

Reply:
As far as I can understand thus fix is trivial.
I'm planning to provide PR in a 24 hour after this PR finish. Is it OK?

Reply:
Wouldn't call it trivial, but sure, this sound good! Thanks a lot. Just want to make sure we get it on time to not miss code freeze deadline. Thanks a lot!